opm/autodiff/BlackoilModelEbos.hpp

```diff
@@ -1045,150 +1046,100 @@ namespace Opm {
                 }
+            }
 
-                hcpv = std::vector<double>(dims, 0.0);
-                pres = std::vector<double>(dims, 0.0);
+            comm.sum(hcpv.data(), hcpv.size());
+            comm.sum(pres.data(), pres.size());
 
-                for (int c = 0; c < nc; ++c) {
-                    const int region = fipnum[c] - 1;
-                    if (region != -1 && mask[c]) {
-                        const auto& intQuants = *ebosSimulator_.model().cachedIntensiveQuantities(c, /*timeIdx=*/0);
-                        const auto& fs = intQuants.fluidState();
-                        const double hydrocarbon = fs.saturation(FluidSystem::oilPhaseIdx).value() + fs.saturation(FluidSystem::gasPhaseIdx).value();
-                        hcpv[region] += pv[c] * hydrocarbon;
-                        pres[region] += pv[c] * fs.pressure(FluidSystem::oilPhaseIdx).value();
-                    }
+            elemIt = elemCtx.gridView().template begin</*codim=*/0>();
+            for (; elemIt != elemEndIt; ++elemIt) {
+                const auto& elem = *elemIt;
+                if (elem.partitionType() != Dune::InteriorEntity) {
+                    continue;
                 }
 
-                comm.sum(hcpv.data(), hcpv.size());
-                comm.sum(pres.data(), pres.size());
-
-                for (int c = 0; c < nc; ++c) {
-                    const int region = fipnum[c] - 1;
-                    if (region != -1 && mask[c]) {
-                        fip_.fip[FIPDataType::FIP_PV][c] = pv[c];
-                        const auto& intQuants = *ebosSimulator_.model().cachedIntensiveQuantities(c, /*timeIdx=*/0);
-                        const auto& fs = intQuants.fluidState();
-                        const double hydrocarbon = fs.saturation(FluidSystem::oilPhaseIdx).value() + fs.saturation(FluidSystem::gasPhaseIdx).value();
-
-                        if (hcpv[region] != 0) {
-                            fip_.fip[FIPDataType::FIP_WEIGHTED_PRESSURE][c] = pv[c] * fs.pressure(FluidSystem::oilPhaseIdx).value() * hydrocarbon / hcpv[region];
-                        } else {
-                            fip_.fip[FIPDataType::FIP_WEIGHTED_PRESSURE][c] = pres[region] / pv[c];
-                        }
+                elemCtx.updatePrimaryStencil(elem);
+                elemCtx.updatePrimaryIntensiveQuantities(/*timeIdx=*/0);
 
-                        values[region][FIPDataType::FIP_PV] += fip_.fip[FIPDataType::FIP_PV][c];
-                        values[region][FIPDataType::FIP_WEIGHTED_PRESSURE] += fip_.fip[FIPDataType::FIP_WEIGHTED_PRESSURE][c];
+                unsigned cellIdx = elemCtx.globalSpaceIndex(/*spaceIdx=*/0, /*timeIdx=*/0);
+                const int region = fipnum[cellIdx] - 1;
+                if (region != -1) {
+                    const auto& intQuants = elemCtx.intensiveQuantities(/*spaceIdx=*/0, /*timeIdx=*/0);
+                    const auto& fs = intQuants.fluidState();
+                    const double pv =
```

Comment:
it is just a single multiplication, so storing it somewhere is not worth the hassle IMO. (also, it would slightly increase memory consumption and it may even be slightly slower due to cache effects; Yeah, we don't need to care about memory consumption and this part of the code is not very performance critical...)